### PR TITLE
support python3 as well as python2

### PR DESCRIPTION
**Pull request by andrew853**
The next release of Access Server (2.9) has migrated to python3 since python2 has reached its end of life. 

Python3 bring some library changes and strict differentiation between bytes and string. 

This fix ensures that the Duo post auth scrip can be used by either Access Server 2.8 which is based on python 2.x or 2.9 and future release based on python 3.x.

The following incompatibilities addressed in python3:
1. httplib has been renamed to  http.client
2. Some api from urllib has been moved to urllib.parse
3. base64 accepts and return bytes
4. the type unicode is not supported.

### CODE REVIEW

File: duo_openvpn_as.py [hunk 1]
```diff
@@ -204,20 +203,36 @@ def sign(ikey, skey, method, host, uri, date, sig_version, params):
...
+
+    return auth
 
 
 def encode_params(params):
     """Returns copy of params with unicode strings utf-8 encoded"""
     new_params = {}
     for key, value in params.items():
-        if isinstance(key, unicode):
+        if not isinstance(key, str):
```
Comment (xdesai) [inline duo_openvpn_as.py:233]: `if isinstance(key, text_type)`

File: duo_openvpn_as.py [hunk 2]
```diff
@@ -204,20 +203,36 @@ def sign(ikey, skey, method, host, uri, date, sig_version, params):
...
             key = key.encode("utf-8")
-        if isinstance(value, unicode):
+        if not isinstance(value, str):
```
Comment (xdesai) [inline duo_openvpn_as.py:235]: `if isinstance(value, text_type)`

File: duo_openvpn_as.py [hunk 3]
```diff
@@ -204,20 +203,36 @@ def sign(ikey, skey, method, host, uri, date, sig_version, params):
     Return basic authorization header line with a Duo Web API signature.
     """
     canonical = canonicalize(method, host, uri, params, date, sig_version)
-    if isinstance(skey, unicode):
+    if not isinstance(skey, str):
```
Comment (xdesai) [inline duo_openvpn_as.py:206]: `if isinstance(skey, text_type)`

File: duo_openvpn_as.py [hunk 4]
```diff
@@ -204,20 +203,36 @@ def sign(ikey, skey, method, host, uri, date, sig_version, params):
...
         skey = skey.encode('utf-8')
-    sig = hmac.new(skey, canonical, hashlib.sha1)
+
+    try:
+        sig = hmac.new(skey, canonical, hashlib.sha1)
+
+    except TypeError:
```
Comment (xdesai) [inline duo_openvpn_as.py:212]: I think this except block can go away if we use the text_type check above

File: duo_openvpn_as.py [hunk 5]
```diff
@@ -204,20 +203,36 @@ def sign(ikey, skey, method, host, uri, date, sig_version, params):
     Return basic authorization header line with a Duo Web API signature.
```
Comment (xdesai) [inline duo_openvpn_as.py:203]: The comments I'm leaving for this function are essentially trying to get it to look more like
https://github.com/duosecurity/duo_client_python/blob/master/duo_client/client.py#L107
just wanted to help you see where I'm coming from. Hopefully this helps!

File: duo_openvpn_as.py [hunk 6]
```diff
@@ -204,20 +203,36 @@ def sign(ikey, skey, method, host, uri, date, sig_version, params):
...
+
+    except TypeError:
+        # python3 require bytes
+        skey = skey.encode('utf-8')
+        sig = hmac.new(skey, canonical.encode(), hashlib.sha1)
+
     auth = '%s:%s' % (ikey, sig.hexdigest())
-    return 'Basic %s' % base64.b64encode(auth)
+
+    try:
```
Comment (xdesai) [inline duo_openvpn_as.py:219]: what you have here seems like it should work, but for consistency I'd prefer if we could mimic what we do in duo_client_python
```
auth = '%s:%s' % (ikey, sig.hexdigest())
if isinstance(auth, text_type):
        auth = auth.encode('utf-8')
b64 = base64.b64encode(auth)
if not isinstance(b64, six.text_type):
    b64 = b64.decode('utf-8')
return 'Basic %s' % b64
```

**Review by xdesai — CHANGES_REQUESTED** [review]
Thanks so much for creating this PR. We appreciate you helping us keep this code up to date. I have a couple spots where I'd like to see some changes with regards to the bytes vs string issue.

How we've handled the bytes vs string issues with python2 and python3 in the past is by using the library `six`. For example https://github.com/duosecurity/duo_client_python/blob/master/duo_client/client.py

Since we don't really have the option for including `six` with the openvpn_as package I think I'd like to see us do our best to emulate its behavior with the imports. Very similar to how you've done it with the httplib and urllib.

For an equivalent to six.text_type the following import should help out.
```
try:
    # Unicode only exists in python 2
    text_type = unicode
except NameError:
    # Python 3 text_type is the string object
    text_type = str
```

File: duo_openvpn_as.py (hunk 1 above)
Comment (andrew853) [reply duo_openvpn_as.py]: done

File: duo_openvpn_as.py (hunk 2 above)
Comment (andrew853) [reply duo_openvpn_as.py]: done

File: duo_openvpn_as.py (hunk 3 above)
Comment (andrew853) [reply duo_openvpn_as.py]: done

File: duo_openvpn_as.py (hunk 4 above)
Comment (andrew853) [reply duo_openvpn_as.py]: done

File: duo_openvpn_as.py (hunk 5 above)
Comment (andrew853) [reply duo_openvpn_as.py:210]: yes, this helps.

File: duo_openvpn_as.py (hunk 6 above)
Comment (andrew853) [reply duo_openvpn_as.py]: All the unicode stuff are changed to the six text_type recommendation.

**Review by xdesai — APPROVED** [review]
This looks good to me!
I'm gonna pull it down and test manually. In the meantime I'll have one of my teammates also look it over.
Thanks for making those changes :)

**Comment by xdesai** [comment]
Got the green check from a couple coworkers. Merging and releasing now.

**Pull request merged**